File: i.sortestpath/boj1753.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

const int INF = 987654321;
int V, E, K;
vector<pair<int, int> > g[20'001];

vector<int> dijkstra(int s, int v_num) {
    vector<int> distance(v_num, INF);
    distance[s] = 0;

    priority_queue< pair<int, int> > pq;
    pq.push({0, s});

    while(!pq.empty()) {
        int cost = -pq.top().first;
        int curV = pq.top().second;
        pq.pop();

        if (distance[curV] < cost) { continue; }

        for (int i = 0; i < (int)g[curV].size(); i++) {
            int n = g[curV][i].first;
            int nd = cost + g[curV][i].second;

            if (distance[n] > nd) {
                distance[n] = nd;
                pq.push({-nd, n});
            }
        }
    }
    return distance;
}

int main() {
    cin.sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> V >> E;
    cin >> K;

    for (int i = 0; i < E; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        g[u].push_back({v, w});
    }
    auto result = dijkstra(K, V+1);
    for (int i = 1; i <= V; i++) {
        if (result[i] == INF) {
            cout << "INF\n";
        } else {
            cout << result[i] << "\n";
        }
    }

}

File: i.sortestpath/boj1753.py
"""
최단경로
"""

import sys
input = sys.stdin.readline

from queue import PriorityQueue


def dijkstra(s, v_num, g):
    dist = [ 987654321 for _ in range(v_num) ]
    dist[s] = 0

    pq = PriorityQueue()
    pq.put([0, s])
    while not pq.empty():
        c = pq.get()
        curV = c[1]
        cost = c[0]

        if dist[curV] < cost: continue
        if not curV in g.keys(): continue
        for v in g[curV]:
            neighbor = v[0]
            neighbor_dist = v[1] + cost
            if dist[neighbor] > neighbor_dist:
                dist[neighbor] = neighbor_dist
                pq.put([neighbor_dist, neighbor])
    return dist

V, E = map(int, input().split())
K = int(input())

g = {}
for _ in range(E):
    u, v, w = map(int, input().split())
    if u in g.keys():
        g[u].append([v, w])
    else:
        g[u] = [[v, w]]

result = dijkstra(K, V+1, g)
print(result)
for r in result[1:]:
    if r == 987654321:
        print("INF")
    else:
        print(r)